### PR TITLE
 Implement ANTLR4 grammar customized linter

### PR DESCRIPTION
**Pull request by jimidle**
Here we implement a linting tool that is customized towards the antlr4 grammars in this repo.

As a first step, we identify any rules defined in a parser grammar that are not referenced anywhere else in the grammar, as well as referenced rules that are undefined. Note that we do not process include file in this PR and so this could result in false negatives if rules in one grammar are referenced in an included grammar. At this point we have no included grammars in our main grammars and hence this is as yet, not an issue.

Note that this PR adds a new github action which eventually will fail the build if it find issues in the grammars. As the new linter finds errors in both our parser grammars, we will need to fix those errors before allowing the lint to fail the build.  

The linter produces output akin to:

```

Issues found in grammar: /Users/jim.idle/databricks/remorph/core/src/main/antlr4/com/databricks/labs/remorph/parsers/tsql/TSqlParser.g4
Orphaned rules (rules defined but never used):
  nouse defined at line 1675:
    nouse
        : DEFAULT_LANGUAGE EQ (id | STRING)

  insertColumnNameList defined at line 3321:
    insertColumnNameList: insertColumnId (COMMA insertColumnId)*
        ;

  entityNameForParallelDw defined at line 3290:
    entityNameForParallelDw: schemaDatabase = id | schema = id DOT objectName = id
        ;

  deprecatedTableHint defined at line 3211:
    deprecatedTableHint: LPAREN tableHint RPAREN
        ;

  createDatabaseScopedCredential defined at line 1665:
    createDatabaseScopedCredential
        : CREATE DATABASE SCOPED CREDENTIAL id WITH IDENTITY EQ STRING (SECRET EQ STRING)?

  udtMethodArguments defined at line 3014:
    udtMethodArguments
        : LPAREN argument += executeVarString (COMMA argument += executeVarString)* RPAREN

  entityNameForAzureDw defined at line 3287:
    entityNameForAzureDw: schema = id | schema = id DOT objectName = id
        ;

  indexValue defined at line 3223:
    indexValue: id | INT
        ;

  sybaseLegacyHint defined at line 3214:
    sybaseLegacyHint: HOLDLOCK | NOHOLDLOCK | READPAST | SHARED
        ;


Issues found in grammar: /Users/jim.idle/databricks/remorph/core/src/main/antlr4/com/databricks/labs/remorph/parsers/snowflake/SnowflakeParser.g4
Orphaned rules (rules defined but never used):
  valuesBuilder defined at line 118:
    valuesBuilder: VALUES L_PAREN exprList R_PAREN (COMMA L_PAREN exprList R_PAREN)?
        ;

  jsonPath defined at line 3145:
    jsonPath: jsonPathElem (DOT jsonPathElem)*
        ;

  bracketExpression defined at line 3276:
    bracketExpression: L_PAREN exprList R_PAREN | L_PAREN subquery R_PAREN
        ;

make: *** [antlr-lint] Error 1

```

### CODE REVIEW

File: antlrlinter/src/main/scala/com/databricks/labs/remorph/antlrlinter/RuleDefinition.scala [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.databricks.labs.remorph.antlrlinter
+
+case class RuleDefinition(lineNo: Integer, ruleName: String, refCount: Integer = 0) {
+  def getLineNo: Integer = lineNo
+  def getRuleName: String = ruleName
+  def getRefCount: Integer = refCount
```
[the remark this answers is not in the document]
Comment (jimidle) [reply antlrlinter/src/main/scala/com/databricks/labs/remorph/antlrlinter/RuleDefinition.scala]: Yup. Auto generated by Copilot - I have removed them in the next push

File: pom.xml [hunk 2]
```diff
@@ -23,6 +23,7 @@
     </license>
   </licenses>
   <modules>
+    <module>antlrlinter</module>
```
[the remark this answers is not in the document]
Comment (jimidle) [reply pom.xml:26]: Hmm - well, it is specifically for anltr grammars, but OK - lets change to linter - can always change it again later

File: antlrlinter/src/main/scala/com/databricks/labs/remorph/antlrlinter/RuleDefinition.scala [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package com.databricks.labs.remorph.antlrlinter
+
+case class RuleDefinition(lineNo: Integer, ruleName: String, refCount: Integer = 0) {
```
[the remark this answers is not in the document]
Comment (jimidle) [reply antlrlinter/src/main/scala/com/databricks/labs/remorph/antlrlinter/RuleDefinition.scala]: No - was a WIP, and not the way to count anyway.

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**